### PR TITLE
Remove all new() constraints from lower Options infrastructure types

### PR DESCRIPTION
**Pull request by NinoFloris**
Remove all new constraints from lower Options infrastructure types and move these to to the service collection extension methods.

 - Due to the open generic service registration  the unconstrained service collection extensions would previously throw an exception during first construction of IOptions<T> (via IOptionsFactory<T>) by the container if T didn't actually have an empty constructor. This is now moved to compile time for these extension methods. (this one)
```cs
services.TryAdd(ServiceDescriptor.Transient(typeof(IOptionsFactory<>), typeof(OptionsFactory<>)));
```
- Removing the `new()` constraints on all the Options types allows for new and creative implementations of IOptionsFactory. This is mainly motivated by cases where classes should always have validation before construction and other cases where there's no empty constructor (by default) like F# records.

This is a minor source level break for invalid `T`s which seems harmless. And removing a generic constraint should be binary compatible, so in terms of impact I'd say this is pretty optimal.

/cc @HaoK

Addresses #1463

@cartermp this is a great usability enhancement for F#. 
It allows for an `ImmutableOptionsFactory` to be supplied by Giraffe or Saturn that handles records and immutable update modifications compared to having to use mutable constructorless pocos.

### CODE REVIEW

**Comment by NinoFloris** [comment]
@eilon @haok how would I go about fixing the CI issue?

>Generated code is not up to date. You might need to regenerate the reference assemblies or project list (see docs/ReferenceAssemblies.md and docs/ReferenceResolution.md)

**Comment by NinoFloris** [comment]
So tests are failing again due to reference assembly issues, yet the files have been regenerated correctly, CI issue?

**Comment by Eilon** [comment]
Adding @aspnet/build

**Comment by dougbu** [comment]
The generated ref assembly project file looks odd.  Was it edited manually?

Then again, it's possible regenerating anywhere but on Windows has some issues. What is your dev platform @NinoFloris?

----
FYI it's not necessary to execute eng/scripts/GenerateReferenceAssemblies.ps1 when only changing a couple of projects. If you're editing the files manually, instead run `dotnet msbuild /t:GenerateReferenceSource` in any project directory where you've changed the public API.

**Comment by NinoFloris** [comment]
@dougbu I'm running on macOS, I don't have powershell there so as a matter of fact only did a single project regenerate. I did manually fix up the TFM == netstandard2.0 condition as that was completely wrong (missing half of the condition and no quotes on the other half) but the extra newlines aren't mine.

File: src/Options/Options/ref/Microsoft.Extensions.Options.netstandard2.0.cs [hunk 1]
```diff
@@ -6,14 +6,14 @@ namespace Microsoft.Extensions.DependencyInjection
     public static partial class OptionsServiceCollectionExtensions
     {
         public static Microsoft.Extensions.DependencyInjection.IServiceCollection AddOptions(this Microsoft.Extensions.DependencyInjection.IServiceCollection services) { throw null; }
-        public static Microsoft.Extensions.Options.OptionsBuilder<TOptions> AddOptions<TOptions>(this Microsoft.Extensions.DependencyInjection.IServiceCollection services) where TOptions : class { throw null; }
-        public static Microsoft.Extensions.Options.OptionsBuilder<TOptions> AddOptions<TOptions>(this Microsoft.Extensions.DependencyInjection.IServiceCollection services, string name) where TOptions : class { throw null; }
-        public static Microsoft.Extensions.DependencyInjection.IServiceCollection ConfigureAll<TOptions>(this Microsoft.Extensions.DependencyInjection.IServiceCollection services, System.Action<TOptions> configureOptions) where TOptions : class { throw null; }
+        public static Microsoft.Extensions.Options.OptionsBuilder<TOptions> AddOptions<TOptions>(this Microsoft.Extensions.DependencyInjection.IServiceCollection services) where TOptions : class, new() { throw null; }
```
Comment (pakrym) [inline src/Options/Options/ref/Microsoft.Extensions.Options.netstandard2.0.cs]: This is a breaking change and I don't think this is required.

**Comment by pakrym** [comment]
I don't think we need to add `new()` constraint on extension methods as it makes things even more limiting than they were. Removing internal constrains on the other hand is good and makes IOptionsFactory slightly more of extensibility point than it is now.

File: src/Options/Options/ref/Microsoft.Extensions.Options.csproj [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (dougbu) [inline src/Options/Options/ref/Microsoft.Extensions.Options.csproj]: Bet if you reverted all changes to this file, the Code Check job would pass. But, I lean toward a lot of this being an unnecessary breaking change (as @pakrym described).

**Comment by HaoK** [comment]
We took https://github.com/aspnet/Extensions/pull/2169 instead

**Comment by NinoFloris** [comment]
These notifications have slipped through, sorry, happy to see you got the fix in

**Pull request closed without merging**